test/extensions/watchdog/profile_action/profile_action_test.cc

```diff
@@ -0,0 +1,357 @@
+
+#include "extensions/watchdog/profile_action/config.h"
+#include "extensions/watchdog/profile_action/profile_action.h"
+
+#include "test/mocks/event/mocks.h"
+#include "test/test_common/environment.h"
+#include "test/test_common/simulated_time_system.h"
+#include "test/test_common/test_time.h"
+#include "test/test_common/utility.h"
+
+#include "absl/strings/substitute.h"
+#include "absl/synchronization/mutex.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Watchdog {
+namespace ProfileAction {
+namespace {
+
+// We want to make sure the action is tested with both simulated time and real
+// time, to ensure that it works in production, and that it works in the context
+// of integration tests which are much easier to control with simulated time.
+enum class TimeSystemType { Real, Simulated };
+
+class ProfileActionTest : public testing::TestWithParam<TimeSystemType> {
+protected:
+  ProfileActionTest()
+      : time_system_(makeTimeSystem()), api_(Api::createApiForTest(*time_system_)),
+        dispatcher_(api_->allocateDispatcher("test")), context_({*api_, *dispatcher_}),
+        thread_(api_->threadFactory().createThread(
+            [this]() -> void { dispatcher_->run(Event::Dispatcher::RunType::RunUntilExit); })),
+        test_path_(generateTestPath()) {}
+
+  // Generates a unique path for a testcase.
+  static std::string generateTestPath() {
+    const ::testing::TestInfo* const test_info =
+        ::testing::UnitTest::GetInstance()->current_test_info();
+    const std::string test_param =
+        GetParam() == TimeSystemType::Real ? "TestRealTimeSystem" : "SimulatedTimeSystem";
+
+    std::string test_path = TestEnvironment::temporaryPath(absl::StrJoin(
+        {test_info->test_suite_name(), test_info->name(), (test_param.c_str())}, "/"));
+    TestEnvironment::createPath(test_path);
+
+    return test_path;
+  }
+
+  // Counts the number of non-empty profiles found within a directory.
+  int countNumberOfProfileInPath(const std::string& path) {
+    int nonempty_profiles_found = 0;
+    Filesystem::Directory directory(path);
+
+    for (const Filesystem::DirectoryEntry& entry : directory) {
+      const std::string full_path = path + "/" + entry.name_;
+
+      // Count if its a non-empty file with the prefix of profiles.
+      if (entry.type_ == Filesystem::FileType::Regular &&
+          absl::StartsWith(entry.name_, "ProfileAction") &&
+          api_->fileSystem().fileSize(full_path) > 0) {
+        nonempty_profiles_found++;
+      }
+    }
+
+    return nonempty_profiles_found;
+  }
+
+  static std::unique_ptr<Event::TestTimeSystem> makeTimeSystem() {
+    if (GetParam() == TimeSystemType::Real) {
+      return std::make_unique<Event::GlobalTimeSystem>();
+    }
+    ASSERT(GetParam() == TimeSystemType::Simulated);
+    return std::make_unique<Event::SimulatedTimeSystem>();
+  }
+
+  void
+  setupAction(envoy::extensions::watchdog::profile_action::v3alpha::ProfileActionConfig& config) {
```

Comment:
Consider inlining this. It will make your tests a little more verbose but since they often access action_ directly, it's helpful to see that being set instead of hiding it behind a method.

Reply:
Done. It more the test body is more complete this way, and this wasn't really hiding much.